In [22]:
import pandas as pd
import numpy as np

#### Displaying Series Dataset

In [3]:
cars= pd.read_csv("E:/Dataset/Data/usedcars.csv")
cars.head()

year model  price  mileage   color transmission
0  2011   SEL  21992     7413  Yellow         AUTO
1  2011   SEL  20995    10926    Gray         AUTO
2  2011   SEL  19995     7351  Silver         AUTO
3  2011   SEL  17809    11613    Gray         AUTO
4  2012    SE  17500     8367   White         AUTO

#### When the data is not with header it is said to be as Series

In [6]:
cars.loc[:,'model']

0      SEL
1      SEL
2      SEL
3      SEL
4       SE
      ... 
145    SES
146     SE
147     SE
148     SE
149     SE
Name: model, Length: 150, dtype: object

#### Below data is with a header hence called as  DataFrame

In [7]:
cars.loc[:,['model']]

model
0     SEL
1     SEL
2     SEL
3     SEL
4      SE
..    ...
145   SES
146    SE
147    SE
148    SE
149    SE

[150 rows x 1 columns]

### Missing Values
#####  1.Identifying Missing Values 
#####  2.Handling Missing Values 

In [ ]:
info(): will tell us Count of non missing values are there

In [8]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   year          150 non-null    int64 
 1   model         150 non-null    object
 2   price         150 non-null    int64 
 3   mileage       150 non-null    int64 
 4   color         150 non-null    object
 5   transmission  150 non-null    object
dtypes: int64(3), object(3)
memory usage: 7.2+ KB


In [19]:
Air = pd.read_csv("E:/Dataset/Data/AirQuality.csv",sep=';',decimal=',')
Air.head(10)

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  10/03/2004  18.00.00     2.6       1360.0     150.0      11.9   
1  10/03/2004  19.00.00     2.0       1292.0     112.0       9.4   
2  10/03/2004  20.00.00     2.2       1402.0      88.0       9.0   
3  10/03/2004  21.00.00     2.2       1376.0      80.0       9.2   
4  10/03/2004  22.00.00     1.6       1272.0      51.0       6.5   
5  10/03/2004  23.00.00     1.2       1197.0      38.0       4.7   
6  11/03/2004  00.00.00     1.2       1185.0      31.0       3.6   
7  11/03/2004  01.00.00     1.0       1136.0      31.0       3.3   
8  11/03/2004  02.00.00     0.9       1094.0      24.0       2.3   
9  11/03/2004  03.00.00     0.6       1010.0      19.0       1.7   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   
5          750.0     89.0        1337.0     96.0        1393.0        949.0   
6          690.0     62.0        1462.0     77.0        1333.0        733.0   
7          672.0     62.0        1453.0     76.0        1333.0        730.0   
8          609.0     45.0        1579.0     60.0        1276.0        620.0   
9          561.0   -200.0        1705.0   -200.0        1235.0        501.0   

      T    RH      AH  Unnamed: 15  Unnamed: 16  
0  13.6  48.9  0.7578          NaN          NaN  
1  13.3  47.7  0.7255          NaN          NaN  
2  11.9  54.0  0.7502          NaN          NaN  
3  11.0  60.0  0.7867          NaN          NaN  
4  11.2  59.6  0.7888          NaN          NaN  
5  11.2  59.2  0.7848          NaN          NaN  
6  11.3  56.8  0.7603          NaN          NaN  
7  10.7  60.0  0.7702          NaN          NaN  
8  10.7  59.7  0.7648          NaN          NaN  
9  10.3  60.2  0.7517          NaN          NaN

In [20]:
Air.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9357 non-null   float64
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(15), object(2)
memory usage: 1.2+ MB


####  We observe that there are last 2 Columns which are completly empty
#### In Pandas by default the following values are treated as NaN: '', '#N/A', '#N/A N/A', '#NA', '-1.#IND', '-1.#QNAN', '-NaN', '-nan', '1.#IND', '1.#QNAN', '<NA>',
####  'NULL','NaN', 'n/a','nan', 'null'. 'N/A', 'NA', 

#### In Pandas if there are any Na values which are not like NaN then we can use na_values=['NAN','naN']

#### Here Nan is represnted as nan but in pandas it is treated as NaN
#### This function helps us to Identify NaN values

#### In Numpy there is a function called as nan

In [23]:
Marks = np.array([40,45,np.nan,60,70])

In [25]:
Marks**2

array([1600., 2025.,   nan, 3600., 4900.])

In [14]:
cars = pd.read_csv("E:/Dataset/Data/usedcars.csv")
cars.head(10)

year model  price  mileage   color transmission
0  2011   SEL  21992     7413  Yellow         AUTO
1  2011   SEL  20995    10926    Gray         AUTO
2  2011   SEL  19995     7351  Silver         AUTO
3  2011   SEL  17809    11613    Gray         AUTO
4  2012    SE  17500     8367   White         AUTO
5  2010   SEL  17495    25125  Silver         AUTO
6  2011   SEL  17000    27393    Blue         AUTO
7  2010   SEL  16995    21026  Silver         AUTO
8  2011   SES  16995    32655  Silver         AUTO
9  2010   SES  16995    36116  Silver         AUTO

In [17]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   year          150 non-null    int64 
 1   model         150 non-null    object
 2   price         150 non-null    int64 
 3   mileage       150 non-null    int64 
 4   color         150 non-null    object
 5   transmission  150 non-null    object
dtypes: int64(3), object(3)
memory usage: 7.2+ KB


In [27]:
Air = pd.read_csv("E:/Dataset/Data/AirQuality.csv",sep=';',decimal=',')
Air.head(10)

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  10/03/2004  18.00.00     2.6       1360.0     150.0      11.9   
1  10/03/2004  19.00.00     2.0       1292.0     112.0       9.4   
2  10/03/2004  20.00.00     2.2       1402.0      88.0       9.0   
3  10/03/2004  21.00.00     2.2       1376.0      80.0       9.2   
4  10/03/2004  22.00.00     1.6       1272.0      51.0       6.5   
5  10/03/2004  23.00.00     1.2       1197.0      38.0       4.7   
6  11/03/2004  00.00.00     1.2       1185.0      31.0       3.6   
7  11/03/2004  01.00.00     1.0       1136.0      31.0       3.3   
8  11/03/2004  02.00.00     0.9       1094.0      24.0       2.3   
9  11/03/2004  03.00.00     0.6       1010.0      19.0       1.7   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   
5          750.0     89.0        1337.0     96.0        1393.0        949.0   
6          690.0     62.0        1462.0     77.0        1333.0        733.0   
7          672.0     62.0        1453.0     76.0        1333.0        730.0   
8          609.0     45.0        1579.0     60.0        1276.0        620.0   
9          561.0   -200.0        1705.0   -200.0        1235.0        501.0   

      T    RH      AH  Unnamed: 15  Unnamed: 16  
0  13.6  48.9  0.7578          NaN          NaN  
1  13.3  47.7  0.7255          NaN          NaN  
2  11.9  54.0  0.7502          NaN          NaN  
3  11.0  60.0  0.7867          NaN          NaN  
4  11.2  59.6  0.7888          NaN          NaN  
5  11.2  59.2  0.7848          NaN          NaN  
6  11.3  56.8  0.7603          NaN          NaN  
7  10.7  60.0  0.7702          NaN          NaN  
8  10.7  59.7  0.7648          NaN          NaN  
9  10.3  60.2  0.7517          NaN          NaN

In [28]:
Air.tail(10)

Date Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  PT08.S2(NMHC)  \
9461  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9462  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9463  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9464  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9465  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9466  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9467  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9468  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9469  NaN  NaN     NaN          NaN       NaN       NaN            NaN   
9470  NaN  NaN     NaN          NaN       NaN       NaN            NaN   

      NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)   T  RH  AH  \
9461      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9462      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9463      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9464      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9465      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9466      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9467      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9468      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9469      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   
9470      NaN           NaN      NaN           NaN          NaN NaN NaN NaN   

      Unnamed: 15  Unnamed: 16  
9461          NaN          NaN  
9462          NaN          NaN  
9463          NaN          NaN  
9464          NaN          NaN  
9465          NaN          NaN  
9466          NaN          NaN  
9467          NaN          NaN  
9468          NaN          NaN  
9469          NaN          NaN  
9470          NaN          NaN

In [29]:
Air.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9357 non-null   float64
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(15), object(2)
memory usage: 1.2+ MB


#### We observe that there are two Columns which are completly empty also there are  
#### Also the Dataset range is upto  9471 in which 9357 values are Not null 114 rows are NaN 

In [30]:
9471-9357

114

#### To check the count of NaN values above we have checked it manually,
#### Instead we can use isna().sum() to find the NaN values and Print the Count of it

In [35]:
Air.isna().sum()

Date              114
Time              114
CO(GT)            114
PT08.S1(CO)       114
NMHC(GT)          114
C6H6(GT)          114
PT08.S2(NMHC)     114
NOx(GT)           114
PT08.S3(NOx)      114
NO2(GT)           114
PT08.S4(NO2)      114
PT08.S5(O3)       114
T                 114
RH                114
AH                114
Unnamed: 15      9471
Unnamed: 16      9471
dtype: int64

#### Checking the Not Null Values

In [36]:
Air.notna().sum()

Date             9357
Time             9357
CO(GT)           9357
PT08.S1(CO)      9357
NMHC(GT)         9357
C6H6(GT)         9357
PT08.S2(NMHC)    9357
NOx(GT)          9357
PT08.S3(NOx)     9357
NO2(GT)          9357
PT08.S4(NO2)     9357
PT08.S5(O3)      9357
T                9357
RH               9357
AH               9357
Unnamed: 15         0
Unnamed: 16         0
dtype: int64

In [37]:
cars.notna().sum()

year            150
model           150
price           150
mileage         150
color           150
transmission    150
dtype: int64

In [38]:
cars.isnull()

year  model  price  mileage  color  transmission
0    False  False  False    False  False         False
1    False  False  False    False  False         False
2    False  False  False    False  False         False
3    False  False  False    False  False         False
4    False  False  False    False  False         False
..     ...    ...    ...      ...    ...           ...
145  False  False  False    False  False         False
146  False  False  False    False  False         False
147  False  False  False    False  False         False
148  False  False  False    False  False         False
149  False  False  False    False  False         False

[150 rows x 6 columns]

In [40]:
Students = pd.read_csv("E:/Dataset/Data/snsdata.csv")
Students.head(10)

gradyear gender     age  friends  basketball  football  soccer  softball  \
0      2006      M     NaN        7           0         0       0         0   
1      2006      F  18.801        0           0         1       0         0   
2      2006      M  18.335       69           0         1       0         0   
3      2006      F  18.875        0           0         0       0         0   
4      2006    NaN  18.995       10           0         0       0         0   
5      2006      F     NaN      142           0         0       0         0   
6      2006      F  18.930       72           0         0       0         0   
7      2006      M  18.322       17           0         0       0         1   
8      2006      F  19.055       52           0         0       0         0   
9      2006      F  18.708       39           0         0       0         0   

   volleyball  swimming  ...  blonde  mall  shopping  clothes  hollister  \
0           0         0  ...       0     0         0        0          0   
1           0         0  ...       0     1         0        0          0   
2           0         0  ...       0     0         0        0          0   
3           0         0  ...       0     0         0        0          0   
4           0         0  ...       0     0         2        0          0   
5           0         0  ...       0     0         1        0          0   
6           0         0  ...       0     2         0        0          2   
7           0         0  ...       0     0         0        0          0   
8           0         0  ...       0     0         0        0          0   
9           0         0  ...       0     0         1        0          0   

   abercrombie  die  death  drunk  drugs  
0            0    0      0      0      0  
1            0    0      0      0      0  
2            0    0      1      0      0  
3            0    0      0      0      0  
4            0    0      0      1      1  
5            0    0      0      1      0  
6            0    0      0      0      0  
7            0    0      0      0      0  
8            0    0      0      0      0  
9            0    0      0      0      0  

[10 rows x 40 columns]

In [42]:
Students.shape

(30000, 40)

In [43]:
Students.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 40 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   gradyear      30000 non-null  int64  
 1   gender        27276 non-null  object 
 2   age           24913 non-null  float64
 3   friends       30000 non-null  int64  
 4   basketball    30000 non-null  int64  
 5   football      30000 non-null  int64  
 6   soccer        30000 non-null  int64  
 7   softball      30000 non-null  int64  
 8   volleyball    30000 non-null  int64  
 9   swimming      30000 non-null  int64  
 10  cheerleading  30000 non-null  int64  
 11  baseball      30000 non-null  int64  
 12  tennis        30000 non-null  int64  
 13  sports        30000 non-null  int64  
 14  cute          30000 non-null  int64  
 15  sex           30000 non-null  int64  
 16  sexy          30000 non-null  int64  
 17  hot           30000 non-null  int64  
 18  kissed        30000 non-nu

In [44]:
Students.isna().sum()

gradyear           0
gender          2724
age             5087
friends            0
basketball         0
football           0
soccer             0
softball           0
volleyball         0
swimming           0
cheerleading       0
baseball           0
tennis             0
sports             0
cute               0
sex                0
sexy               0
hot                0
kissed             0
dance              0
band               0
marching           0
music              0
rock               0
god                0
church             0
jesus              0
bible              0
hair               0
dress              0
blonde             0
mall               0
shopping           0
clothes            0
hollister          0
abercrombie        0
die                0
death              0
drunk              0
drugs              0
dtype: int64

In [45]:
cars.isna().sum()

year            0
model           0
price           0
mileage         0
color           0
transmission    0
dtype: int64

In [46]:
cars.transmission.value_counts()

AUTO      128
MANUAL     22
Name: transmission, dtype: int64

In [48]:
Students.gender.value_counts()

F    22054
M     5222
Name: gender, dtype: int64